client/components/date-range/index.js

```diff
@@ -0,0 +1,665 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import { noop, isNil, isNull, has } from 'lodash';
+import { DateUtils } from 'react-day-picker';
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { localize, moment } from 'i18n-calypso';
+import DatePicker from 'components/date-picker';
+import Popover from 'components/popover';
+import DateRangeInputs from './inputs';
+import DateRangeHeader from './header';
+import DateRangeTrigger from './trigger';
+import Button from 'components/button';
+import Gridicon from 'gridicons';
+const NO_DATE_SELECTED_VALUE = null;
+
+export class DateRange extends Component {
+	static propTypes = {
+		selectedStartDate: PropTypes.oneOfType( [
+			PropTypes.instanceOf( Date ),
+			PropTypes.instanceOf( moment ),
+		] ),
+		selectedEndDate: PropTypes.oneOfType( [
+			PropTypes.instanceOf( Date ),
+			PropTypes.instanceOf( moment ),
+		] ),
+		onDateSelect: PropTypes.func,
+		onDateCommit: PropTypes.func,
+		firstSelectableDate: PropTypes.oneOfType( [
+			PropTypes.instanceOf( Date ),
+			PropTypes.instanceOf( moment ),
+		] ),
+		lastSelectableDate: PropTypes.oneOfType( [
+			PropTypes.instanceOf( Date ),
+			PropTypes.instanceOf( moment ),
+		] ),
+		triggerText: PropTypes.func,
+		isCompact: PropTypes.bool,
+		renderTrigger: PropTypes.func,
+		renderHeader: PropTypes.func,
+		renderInputs: PropTypes.func,
+	};
+
+	static defaultProps = {
+		onDateSelect: noop,
+		onDateCommit: noop,
+		isCompact: false,
+		focusedMonth: null,
+		renderTrigger: props => <DateRangeTrigger { ...props } />,
+		renderHeader: props => <DateRangeHeader { ...props } />,
+		renderInputs: props => <DateRangeInputs { ...props } />,
+	};
+
+	constructor( props ) {
+		super( props );
+
+		// Define the date range as Moment instances
+		const firstSelectableDate =
+			has( this.props, 'firstSelectableDate' ) &&
+			this.props.moment( this.props.firstSelectableDate );
+		const lastSelectableDate =
+			has( this.props, 'lastSelectableDate' ) && this.props.moment( this.props.lastSelectableDate );
+
+		// Clamp dates to ranges (if specified)
+		let startDate;
+		let endDate;
+
+		endDate = isNil( this.props.selectedEndDate )
+			? NO_DATE_SELECTED_VALUE
+			: this.props.moment( this.props.selectedEndDate );
+
+		if ( ! isNull( endDate ) ) {
+			endDate = this.clampDateToRange( endDate, {
+				dateFrom: firstSelectableDate,
+				dateTo: lastSelectableDate,
+			} );
+		}
+
+		startDate = isNil( this.props.selectedStartDate )
```

Comment:
This seems to be more complicated than it need be, with some indirection that might be avoidable:

For starters, here are some things to consider:
```js
// isNil === null or undefined.
// isNull === null.
// NO_DATE_SELECTED_VALUE = null;
```
so 
```js
endDate = isNil( this.props.selectedEndDate )
	? NO_DATE_SELECTED_VALUE
	: this.props.moment( this.props.selectedEndDate );

// essentially does this:
if ( selectedEndDate === null ) {
  endDate = null;
} else if ( selectedEndDate === undefined ) {
  endDate = null;
} else {
  endDate = this.props.moment( this.props.selectedEndDate )
}
```
We then check for `! isNull( endDate )` but we already know that it's either `null` or a `moment` because we coerced any `undefined`s to `NO_DATE_SELECTED_VALUE`/`null`

Because of that, I think we could merge the two conditionals in to:

```js
const endDate = isNil( selectedEndDate )
	? NO_DATE_SELECTED_VALUE
	: this.clampDateToRange(
		this.props.moment( selectedEndDate ), 
		{
			dateFrom: firstSelectableDate,
			dateTo: lastSelectableDate,
		}
	);
```

Reply:
With the above we could make a couple of extra tweaks that would allow us to avoid using `let` for `endDate` and `startDate`, though I'm not suggesting we do (not worth the time).
We could theoretically do so by making and using a util like so 
```js
startDate: getEarliest( startDate, endDate ),
endDate: getLatest( startDate, endDate ),
initialStartDate: getEarliest( startDate, endDate ),
initialEndDate: getLatest( startDate, endDate ),
```

Reply:
Thanks. I never liked this but was focusing attention elsewhere.

Have simplified to a degree with the MVP and time restrictions in mind.

Reply:
Awesome, thanks @getdave !